src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -1337,16 +1423,70 @@ int computeIndexingBin() {
+     *
+     * A SAMRecord with a null header can not be fully validated and will result in SAMValidationErrors if isValid is
+     * called.
+     *
+     * NOTE: SAMTextWriter, BAMFileWriter, and CRAMFileWriter all require records to have a valid header in order to be
+     * written. Any record that does not have a header at the time it is added to the writer will be updated to use the
+     * header associated with the writer. Additionally, record comparators (i.e. SAMRecordCoordinateComparator and
+     * SAMRecordDuplicateComparator) require records with non-null header values.
+     */
     public SAMFileHeader getHeader() {
         return mHeader;
     }
 
     /**
-     * Setting header into SAMRecord facilitates conversion btw reference sequence names and indices
+     * Sets the SAMFileHeader for this record. Setting the header into SAMRecord facilitates conversion between reference
+     * sequence names and indices.
+     *
+     * If the record has a reference or mate reference name, the corresponding reference and mate reference indices are
+     * resolved and updated using the sequence dictionary in the new header.
+     *
+     * When the SAMFileHeader is set to null, the reference and mate reference indices are cleared. Therefore, calls to
+     * the following SAMRecord methods on records with a null header may throw IllegalArgumentExceptions:
+     *
+     *   getReferenceIndex
+     *   setReferenceIndex
+     *   getMateReferenceIndex
+     *   setMateReferenceIndex
+     *
+     * A SAMRecord with a null header can not be fully validated and will result in SAMValidationErrors if isValid is
+     * called.
+     *
+     * NOTE: SAMTextWriter, BAMFileWriter, and CRAMFileWriter all require records to have a valid header in order to be
+     * written. Any record that does not have a header at the time it is added to the writer will be updated to use the
+     * header associated with the writer. Additionally, record comparators (i.e. SAMRecordCoordinateComparator and
+     * SAMRecordDuplicateComparator) require records with non-null header values.
+     *
      * @param header contains sequence dictionary for this SAMRecord
      */
     public void setHeader(final SAMFileHeader header) {
         this.mHeader = header;
+        if (null == header) {
+            // if the reference and mate reference index values have previously been resolved, set them to unresolved
+            if (null != mReferenceIndex && !mReferenceIndex.equals(NO_ALIGNMENT_REFERENCE_INDEX)) {
+                mReferenceIndex = null;
+            }
+            if (null != mMateReferenceIndex && !mMateReferenceIndex.equals(NO_ALIGNMENT_REFERENCE_INDEX)) {
+                mMateReferenceIndex = null;
+            }
+        }
+        else {
+            // attempt to resolve the existing reference names and indices against the new sequence dictionary, but
+            // don't throw if the names don't appear in the dictionary
```

Comment:
Make this a prominent warning in the method-level docs (that we won't throw if the existing reference names aren't present in the new header)